File: 007 有序表的合并/使用链表实现.swift
import Foundation

func merge(_ listA: LinkList<Int>, from listB: LinkList<Int>) {
    var pA: Node<Int>? = listA.header
    let pB: Node<Int>? = listB.header
    
    var nA = pA?.next
    var nB = pB?.next
    
    while nA != nil && nB != nil {
        if nA!.value! < nB!.value! {
            pA = nA
            nA = pA?.next
        } else {
            pB?.next = nB?.next
            
            pA?.next = nB
            nB?.next = nA
            pA = nB
            
            nB = pB?.next
        }
    }
    
    if let nB = nB {
        pA?.next = nB
    }
}

func merge(_ listA: LinkList<Int>, _ listB: LinkList<Int>) -> LinkList<Int> {
    var pA: Node<Int>? = listA.header.next
    var pB: Node<Int>? = listB.header.next
    
    let listC = LinkList<Int>()
    var pC: Node<Int>? = listC.header
    
    while pA != nil && pB != nil {
        if pA!.value! < pB!.value! {
            pC?.next = pA
            pA = pA?.next
        } else {
            pC?.next = pB
            pB = pB?.next
        }
        pC = pC?.next
    }
    
    if let pB = pB {
        pC?.next = pB
    }
    
    return listC
}

// 测试数据
var arrA = [Int]()
var arrB = [Int]()

for _ in 0..<6 {
    arrA.append(Int.random(in: 0..<40))
}

for _ in 0..<9 {
    arrB.append(Int.random(in: 0..<90))
}

let listA = LinkList<Int>()
let listB = LinkList<Int>()

for v in arrA.sorted() {
    listA.append(v)
}

for v in arrB.sorted() {
    listB.append(v)
}


print("数组A：")
listA.traverse()
print("数组B：")
listB.traverse()

//print("合并后数组：")
//merge(listA, from: listB)
//listA.traverse()

print("合并后数组：")
let listC = merge(listA, listB)
listC.traverse()

File: 007 有序表的合并/使用链表实现.c
#include <stdio.h>
#include <stdlib.h>

// 合并到一个新的链表中
void merge(LinkList *listA, LinkList *listB, LinkList *listC) {
    Node *pA = listA->next;
    Node *pB = listB->next;
    Node *pC = listC;
    
    while (pA != NULL && pB != NULL) {
        if (pA->value < pB->value) {
            pC->next = pA;
            pA = pA->next;
        } else {
            pC->next = pB;
            pB = pB->next;
        }
        pC = pC->next;
    }
    
    if (pA != NULL) {
        pC->next = pA;
    }
    
    if (pB != NULL) {
        pC->next = pB;
    }
}

// 合并到一个链表 listA 中
void merge2(LinkList *listA, LinkList *listB) {
    Node *pA = listA;
    Node *pB = listB;
    
    Node *nA = pA->next;
    Node *nB = pB->next;
    
    while (nA != NULL && nB != NULL) {
        
        if (nA->value < nB->value) {
            pA = nA;
            nA = pA->next;
        } else {
            pB->next = nB->next;
            
            pA->next = nB;
            nB->next = nA;
            pA = nB;
            
            nB = pB->next;
        }
    }
    
    if (nB != NULL) {
        pA->next = nB;
    }
}

int main() {
    int arrA[4] = {21, 32, 36, 88};
    int arrB[7] = {2, 15, 37, 99, 123, 345, 432};
    
    
    LinkList *listA = createList();
    LinkList *listB = createList();
    LinkList *listC = createList();
    
    for (int i = 0; i < 4; i++) {
        append(listA, arrA[i]);
    }
    
    for (int i = 0; i < 7; i++) {
        append(listB, arrB[i]);
    }
    
    printf("数组A：");
    traverse(listA);
    printf("数组B：");
    traverse(listB);
    
    merge(listA, listB, listC);
    printf("合并后数组：");
    traverse(listC);
    
//    merge2(listA, listB);
//    printf("合并后数组：");
//    traverse(listA);
    
    return 0;
}
